init.lua

```diff
@@ -1,116 +1,540 @@
+			.. '" has stopped looking over your shoulder.')
+
+	end
+	core_log('action', '[spectator_mode] "' .. name_watcher
+		.. '" detached from "' .. state.target .. '"')
+
+end -- detach
 
-minetest.register_chatcommand("watch", {
-	params = "<to_name>",
-	description = "Watch a given player",
-	privs = {watch = true},
-	func = function(name_watcher, name_target)
+
+-- both players are online and all checks have been done when this
+-- method is called
+local function attach(name_watcher, name_target)
+
+	-- detach from cart, horse, bike etc.
+	detach(name_watcher)
+
+	local watcher = get_player_by_name(name_watcher)
+	local privs_watcher = get_player_privs(name_watcher)
+	-- back up some attributes
+	local properties = watcher:get_properties()
+	local state = {
+		collisionbox = properties.collisionbox,
+		hud_flags = watcher:hud_get_flags(),
+		makes_footstep_sound = properties.makes_footstep_sound,
+		nametag = watcher:get_nametag_attributes(),
+		pos = watcher:get_pos(),
+		priv_interact = privs_watcher.interact,
+		privs_extra = {},
+		target = name_target,
+		visual_size = properties.visual_size,
+	}
+	local privs_extra = invites[name_watcher] and sm.extra_observe_privs
+		or sm.extra_observe_privs_moderator
+
+	for key, _ in pairs(privs_extra) do
+		state.privs_extra[key] = privs_watcher[key]
+		privs_watcher[key] = true
+	end
+	original_state_set(watcher, state)
+
+	-- set some attributes
+	turn_off_hud_flags(watcher)
+	watcher:set_properties({
+		visual_size = { x = 0, y = 0 },
+		makes_footstep_sound = false,
+		collisionbox = { 0 }, -- TODO: is this the proper/best way?
+	})
+	watcher:set_nametag_attributes({ color = { a = 0 }, bgcolor = { a = 0 } })
+	local eye_pos = vector_new(0, -5, -20)
+	watcher:set_eye_offset(eye_pos)
+	-- make sure watcher can't interact
+	privs_watcher.interact = nil
+	set_player_privs(name_watcher, privs_watcher)
+	-- and attach
+	player_api.player_attached[name_watcher] = true
+	local target = get_player_by_name(name_target)
+	watcher:set_attach(target, '', eye_pos)
+	core_log('action', '[spectator_mode] "' .. name_watcher
+		.. '" attached to "' .. name_target .. '"')
+
+	if sm.keep_all_observers_alive or (not invites[name_watcher]) then
+		-- server keeps all observers alive
+		-- or moderator used '/watch' to sneak up without invite
+		after(3, sm.keep_alive, name_watcher)
+	end
+end -- attach
+
+
+-- called by '/watch' command
+local function watch(name_watcher, name_target)
+	if original_state[name_watcher] then
+		return true, 'You are currently watching "'
+			.. original_state[name_watcher].target
+			.. '". Say /' .. sm.command_detach .. ' first.'
+
+	end
+	if name_watcher == name_target then
+		return true, 'You may not watch yourself.'
+	end
+
+	local target = get_player_by_name(name_target)
+	if not target then
+		return true, 'Invalid target name "' .. name_target .. '"'
+	end
+
+	-- avoid infinite loops
+	-- TODO: should we just watch the watched one then? Griefers can be a nuisance both ways.
```

Comment:
The current approach is reasonable. What's the deal with griefers? They cannot interact while watching.

Reply:
yes, they can't damage nodes. They can make it hard on moderators to watch them.
Also they can craft and drop items - not that it matters

The comment basically asks if current implementation is good enough. Sounds so. Have removed comment.